{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "8a28aad4",
   "metadata": {},
   "source": [
    "# Tutorial 4: Analysis of large-scale Stereo-seq MOBS1 data"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "5fbdbf4d",
   "metadata": {},
   "source": [
    "The following tutorials demonstrates how to use PAST to derive latent embeddings and decipher spatial domains on Stereo-seq mouse olfactory bulb section1 (MOBS1) dataset([Chen, et al., 2022](https://www.sciencedirect.com/science/article/pii/S0092867422003993)). MOBS1 dataset, containing 107,416 spots and 26,145 genes, is too large to analyze with common ST methods. Here we demonstrate PAST could be applied to MOBS1 data with low time and memory cost.\n",
    "\n",
    "There are two parts in this tutorial:\n",
    "\n",
    "+ **Integrating self-prior data to analyze MOBS1 data.** This part will show you how to utilize target preprocessed gene expression matrix as self-prior matrix to obtain latent embeddings and spatial clusters on Stereo-seq MOBS1 data.\n",
    "\n",
    "+ **Integrating external-prior data to analyze MOBS1 data.** This part will show you how to utilize another mouse olfactory bulb section also generated by Stereo-seq to construct external-prior pseudo-bulk matrix for latent feature extraction and spatial clustering on Stereo-seq MOBS1 data."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "4cd24de2",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2022-10-10T13:37:17.678171Z",
     "start_time": "2022-10-10T13:37:16.521521Z"
    }
   },
   "outputs": [],
   "source": [
    "import past\n",
    "import os\n",
    "import scanpy as sc\n",
    "import warnings\n",
    "import torch\n",
    "import numpy as np\n",
    "import pandas as pd"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "6835e1cc",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2022-10-10T13:37:17.720748Z",
     "start_time": "2022-10-10T13:37:17.679460Z"
    }
   },
   "outputs": [],
   "source": [
    "warnings.filterwarnings(\"ignore\")\n",
    "sc.set_figure_params(dpi=80, figsize=(3,3), facecolor=\"white\")\n",
    "os.environ[\"R_HOME\"] = \"/home/lizhen/miniconda3/envs/scTECH-R/lib/R\"\n",
    "device = torch.device(\"cuda\") if torch.cuda.is_available() else torch.device(\"cpu\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "de43c90e",
   "metadata": {},
   "source": [
    "Using memory_profiler to record time and memory cost of PAST for training and prediction"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "26242aa9",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2022-10-10T13:37:17.727425Z",
     "start_time": "2022-10-10T13:37:17.722281Z"
    }
   },
   "outputs": [],
   "source": [
    "%load_ext memory_profiler"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "615bcbc5",
   "metadata": {},
   "source": [
    "You can uncomment and execute the following command to download the Stereo-seq MOBS1 dataset in AnnData format."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "001aa3b8",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2022-10-10T13:37:17.730922Z",
     "start_time": "2022-10-10T13:37:17.729077Z"
    }
   },
   "outputs": [],
   "source": [
    "# !wget https://health.tsinghua.edu.cn/software/PAST/data/Stereo-seq_MOBS1.h5ad"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "e27f5497",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2022-10-10T13:37:17.923586Z",
     "start_time": "2022-10-10T13:37:17.731738Z"
    }
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "AnnData object with n_obs × n_vars = 107416 × 26145\n",
       "    obs: 'annotation', 'n_genes_by_counts', 'log1p_n_genes_by_counts', 'total_counts', 'log1p_total_counts'\n",
       "    var: 'n_cells', 'n_cells_by_counts', 'mean_counts', 'log1p_mean_counts', 'pct_dropout_by_counts', 'total_counts', 'log1p_total_counts'\n",
       "    obsm: 'spatial'"
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "## DLPFC DATASET\n",
    "os.chdir(\"/home/lizhen/code/PAST/Data\")\n",
    "MOBS1 = sc.read_h5ad(filename=\"Stereo-seq_MOBS1.h5ad\")\n",
    "MOBS1.var_names_make_unique()\n",
    "MOBS1"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "f94fe98a",
   "metadata": {},
   "source": [
    "## Integrating self-prior matrix to analyze MOBS1"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "47a7182e",
   "metadata": {},
   "source": [
    "We set a random seed for all random process for reproducibility."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "4ea1c8f7",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2022-10-10T13:37:17.994684Z",
     "start_time": "2022-10-10T13:37:17.924501Z"
    }
   },
   "outputs": [],
   "source": [
    "past.setup_seed(666)\n",
    "sdata = MOBS1.copy()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "e4c217f3",
   "metadata": {},
   "source": [
    "Next, we filter out genes expressed in less than three spots, follow [the Scanpy vignette](https://scanpy-tutorials.readthedocs.io/en/latest/spatial/basic-analysis.html) for spatial transcriptomics to normalize and logarithmize the count matrix and then select the top 3000 spatial variable genes based on the [Geary'C Index](https://www.jstor.org/stable/2986645)."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f680fb00",
   "metadata": {
    "ExecuteTime": {
     "start_time": "2022-10-10T13:37:09.356Z"
    }
   },
   "outputs": [],
   "source": [
    "sdata = past.preprocess(sdata, min_cells=3, is_filter_MT=True, n_tops=3000, gene_method=\"gearyc\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "a343d116",
   "metadata": {},
   "source": [
    "We initialize PAST model with `d_lat=50`, `k_neighbors=6`."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f430663e",
   "metadata": {
    "ExecuteTime": {
     "start_time": "2022-10-10T13:37:09.357Z"
    }
   },
   "outputs": [],
   "source": [
    "PAST_model = past.PAST(d_in=sdata.shape[1], d_lat=50, k_neighbors=6, dropout=0.1).to(device)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "f6d8e6ed",
   "metadata": {},
   "source": [
    "Then we train PAST model without specifying external reference data, and PAST will automatically utilize the preprocessed target MOBS1 anndata as self-prior marix."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "38595f03",
   "metadata": {
    "ExecuteTime": {
     "start_time": "2022-10-10T13:37:09.358Z"
    }
   },
   "outputs": [],
   "source": [
    "%time %memit PAST_model.model_train(sdata, epochs=50, lr=1e-3, batchsize=3600, device=device)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "ea226e55",
   "metadata": {},
   "source": [
    "We can obtain the latent embedding through `output()` function of PAST object, and the latent embedding is stored in `sdata.obsm[\"embedding\"]`."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "644c1dfb",
   "metadata": {
    "ExecuteTime": {
     "start_time": "2022-10-10T13:37:09.359Z"
    }
   },
   "outputs": [],
   "source": [
    "%time %memit sdata = PAST_model.output(sdata)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "baf5ea15",
   "metadata": {},
   "source": [
    "The result of mclust and louvain with default parameters is stored in `sdata.obs[\"mclust\"]` and `sdata.obs[\"Dlouvain\"]` respectively."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "1da5955e",
   "metadata": {
    "ExecuteTime": {
     "start_time": "2022-10-10T13:37:09.360Z"
    }
   },
   "outputs": [],
   "source": [
    "sdata = past.mclust_R(sdata, num_cluster=sdata.obs[\"annotation\"].nunique(), used_obsm='embedding')\n",
    "sdata = past.default_louvain(sdata, use_rep=\"embedding\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "4bf6e9ed",
   "metadata": {},
   "source": [
    "Evaluation of latent embedding and spatial clustering results."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "7b61c02d",
   "metadata": {
    "ExecuteTime": {
     "start_time": "2022-10-10T13:37:09.361Z"
    }
   },
   "outputs": [],
   "source": [
    "print(\"Cross-validation score:\", end=\"\\t\\t\")\n",
    "acc, kappa, mf1, wf1 = past.svm_cross_validation(sdata.obsm[\"embedding\"], sdata.obs[\"annotation\"])\n",
    "print(\"Acc: %.3f, K: %.3f, mF1: %.3f, wF1: %.3f\"%(acc.mean(), kappa.mean(), mf1.mean(), wf1.mean()))\n",
    "\n",
    "print(\"Mclust metrics:\", end=\"\\t\\t\\t\")\n",
    "ari, ami, nmi, fmi, comp, homo = past.cluster_metrics(sdata, \"annotation\", \"mclust\")\n",
    "print(\"ARI: %.3f, AMI: %.3f, NMI: %.3f, FMI:%.3f, Comp: %.3f, Homo: %.3f\"%(ari, ami, nmi, fmi, comp, homo))\n",
    "\n",
    "print(\"Dlouvain metrics:\", end=\"\\t\\t\")\n",
    "ari, ami, nmi, fmi, comp, homo = past.cluster_metrics(sdata, \"annotation\", \"Dlouvain\")\n",
    "print(\"ARI: %.3f, AMI: %.3f, NMI: %.3f, FMI:%.3f, Comp: %.3f, Homo: %.3f\"%(ari, ami, nmi, fmi, comp, homo))"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "1ba15a1c",
   "metadata": {},
   "source": [
    "Visualization of latent embeddings colored by annotation and PAST-derived clusters."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "8453836a",
   "metadata": {
    "ExecuteTime": {
     "start_time": "2022-10-10T13:37:09.362Z"
    },
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "sc.pp.neighbors(sdata, use_rep='embedding')\n",
    "sc.tl.umap(sdata)\n",
    "sc.pl.umap(sdata, color=[\"annotation\", \"mclust\"], title=[\"Ground Truth\", \"PAST-S\"])"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "d386b11b",
   "metadata": {},
   "source": [
    "Visualization of spatial domains colored by annotation and PAST-derived clusters."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "86b70803",
   "metadata": {
    "ExecuteTime": {
     "start_time": "2022-10-10T13:37:09.363Z"
    }
   },
   "outputs": [],
   "source": [
    "sc.pl.embedding(sdata, color=[\"annotation\", \"mclust\"], basis=\"spatial\", title=[\"Ground Truth\", \"PAST-S\"])"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "670a1d5e",
   "metadata": {},
   "source": [
    "## Integrating external-prior matrix to analyze MOBS1"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "95b11f1b",
   "metadata": {},
   "source": [
    "Similarly, We set a random seed for all random process for reproducibility."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "07a31b68",
   "metadata": {
    "ExecuteTime": {
     "start_time": "2022-10-10T13:37:09.364Z"
    }
   },
   "outputs": [],
   "source": [
    "past.setup_seed(666)\n",
    "sdata = MOBS1.copy()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "f78a63d9",
   "metadata": {},
   "source": [
    "As done above, We also filter out genes expressed in less than three spots, follow [the Scanpy vignette](https://scanpy-tutorials.readthedocs.io/en/latest/spatial/basic-analysis.html) for spatial transcriptomics to normalize and logarithmize the count matrix and then select the top 3000 spatial variable genes based on the [Geary'C Index](https://www.jstor.org/stable/2986645)."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "92839b46",
   "metadata": {
    "ExecuteTime": {
     "start_time": "2022-10-10T13:37:09.365Z"
    }
   },
   "outputs": [],
   "source": [
    "sdata = past.preprocess(sdata, min_cells=3, is_filter_MT=True, n_tops=3000, gene_method=\"gearyc\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "74eb43b0",
   "metadata": {},
   "source": [
    "Here we utilize another mouse olfactory bulb section2 (MOBS2) dataset([Chen, et al., 2022](https://www.sciencedirect.com/science/article/pii/S0092867422003993)) as reference data, denoted as `rdata`."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f73b98bc",
   "metadata": {
    "ExecuteTime": {
     "start_time": "2022-10-10T13:37:09.366Z"
    }
   },
   "outputs": [],
   "source": [
    "os.chdir(\"/home/lizhen/code/PAST/Data\")\n",
    "rdata = sc.read_h5ad(filename=\"Stereo-seq_MOBS2.h5ad\")\n",
    "rdata = past.integration(rdata, sdata)\n",
    "rdata = past.preprocess(rdata, min_cells=None, target_sum=None, is_filter_MT=False, n_tops=None)\n",
    "rdata = past.get_bulk(rdata, key=\"annotation\")\n",
    "rdata"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "751630bb",
   "metadata": {},
   "source": [
    "We initialize PAST model with the same parameters."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d6ce3064",
   "metadata": {
    "ExecuteTime": {
     "start_time": "2022-10-10T13:37:09.367Z"
    }
   },
   "outputs": [],
   "source": [
    "PAST_model = past.PAST(d_in=sdata.shape[1], d_lat=50, k_neighbors=6, dropout=0.1).to(device)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "35d4c468",
   "metadata": {},
   "source": [
    "Then we use `rdata` to construct external-prior matrix, and `rdata` should contain raw count gene expression matrix of tissues related to the target data. Besides, users need to specify the `anno_key` stored in `rdata.obs` because the construction of external-prior pseudo-bulk matrix also requires fine annotation of each spot."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "848ccab5",
   "metadata": {
    "ExecuteTime": {
     "start_time": "2022-10-10T13:37:09.368Z"
    }
   },
   "outputs": [],
   "source": [
    "%time %memit PAST_model.model_train(sdata, rdata=rdata, epochs=50, lr=1e-3, batchsize=3600, device=device)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "00e105aa",
   "metadata": {},
   "source": [
    "We then obtain the latent embedding through `output()` function of PAST object, and the latent embedding is stored in `sdata.obsm[\"embedding\"]`."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "fb2ba57c",
   "metadata": {
    "ExecuteTime": {
     "start_time": "2022-10-10T13:37:09.369Z"
    }
   },
   "outputs": [],
   "source": [
    "%time %memit sdata = PAST_model.output(sdata)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "d4094cc6",
   "metadata": {},
   "source": [
    "The result of mclust and louvain with default parameters is stored in `sdata.obs[\"mclust\"]` and `sdata.obs[\"Dlouvain\"]` respectively."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "22c0a175",
   "metadata": {
    "ExecuteTime": {
     "start_time": "2022-10-10T13:37:09.370Z"
    }
   },
   "outputs": [],
   "source": [
    "sdata = past.mclust_R(sdata, num_cluster=sdata.obs[\"annotation\"].nunique(), used_obsm='embedding')\n",
    "sdata = past.default_louvain(sdata, use_rep=\"embedding\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "4ef2b750",
   "metadata": {},
   "source": [
    "Evaluation of latent embedding and spatial clustering results."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c9cbf4ae",
   "metadata": {
    "ExecuteTime": {
     "start_time": "2022-10-10T13:37:09.372Z"
    }
   },
   "outputs": [],
   "source": [
    "print(\"Cross-validation score:\", end=\"\\t\\t\")\n",
    "acc, kappa, mf1, wf1 = past.svm_cross_validation(sdata.obsm[\"embedding\"], sdata.obs[\"annotation\"])\n",
    "print(\"Acc: %.3f, K: %.3f, mF1: %.3f, wF1: %.3f\"%(acc.mean(), kappa.mean(), mf1.mean(), wf1.mean()))\n",
    "\n",
    "print(\"Mclust metrics:\", end=\"\\t\\t\\t\")\n",
    "ari, ami, nmi, fmi, comp, homo = past.cluster_metrics(sdata, \"annotation\", \"mclust\")\n",
    "print(\"ARI: %.3f, AMI: %.3f, NMI: %.3f, FMI:%.3f, Comp: %.3f, Homo: %.3f\"%(ari, ami, nmi, fmi, comp, homo))\n",
    "\n",
    "print(\"Dlouvain metrics:\", end=\"\\t\\t\")\n",
    "ari, ami, nmi, fmi, comp, homo = past.cluster_metrics(sdata, \"annotation\", \"Dlouvain\")\n",
    "print(\"ARI: %.3f, AMI: %.3f, NMI: %.3f, FMI:%.3f, Comp: %.3f, Homo: %.3f\"%(ari, ami, nmi, fmi, comp, homo))"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "459612be",
   "metadata": {},
   "source": [
    "Visualization of latent embeddings colored by annotation and PAST-derived clusters."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "36dda903",
   "metadata": {
    "ExecuteTime": {
     "start_time": "2022-10-10T13:37:09.373Z"
    },
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "sc.pp.neighbors(sdata, use_rep='embedding')\n",
    "sc.tl.umap(sdata)\n",
    "sc.pl.umap(sdata, color=[\"annotation\", \"mclust\"], title=[\"Ground Truth\", \"PAST-E\"])"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "2bafe5eb",
   "metadata": {},
   "source": [
    "Visualization of spatial domains colored by annotation and PAST-derived clusters."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "265c2687",
   "metadata": {
    "ExecuteTime": {
     "start_time": "2022-10-10T13:37:09.374Z"
    }
   },
   "outputs": [],
   "source": [
    "sc.pl.embedding(sdata, color=[\"annotation\", \"mclust\"], basis=\"spatial\", title=[\"Ground Truth\", \"PAST-E\"])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "67e5d5ce",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "torch1.9-py",
   "language": "python",
   "name": "torch1.9"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.12"
  },
  "toc": {
   "base_numbering": 1,
   "nav_menu": {},
   "number_sections": true,
   "sideBar": true,
   "skip_h1_title": false,
   "title_cell": "Table of Contents",
   "title_sidebar": "Contents",
   "toc_cell": false,
   "toc_position": {
    "height": "927.846px",
    "left": "407px",
    "top": "91.97px",
    "width": "314.609px"
   },
   "toc_section_display": true,
   "toc_window_display": true
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
